pymc/distributions/distribution.py

```diff
@@ -473,123 +475,447 @@ def rng_fn(cls, rng, *args):
+
+        @_logcdf.register(rv_type)
+        def density_dist_logcdf(op, value, rng, size, dtype, *dist_params, **kwargs):
+            return logcdf(value, *dist_params, **kwargs)
+
+        @_moment.register(rv_type)
+        def density_dist_get_moment(op, rv, rng, size, dtype, *dist_params):
+            return moment(rv, size, *dist_params)
+
+        rv_op = rv_type()
+        return rv_op(*dist_params, **kwargs)
+
+
+class CustomSymbolicDistRV(SymbolicRandomVariable):
+    """
+    Base class for CustomSymbolicDist
+
+    This should be subclassed when defining custom CustomDist objects that have
+    symbolic random methods.
+    """
+
+    default_output = -1
+
+    _print_name = ("CustomSymbolicDist", "\\operatorname{CustomSymbolicDist}")
+
+    def update(self, node: Node):
+        op = node.op
+        inner_updates = collect_default_updates(op.inner_inputs, op.inner_outputs)
+
+        # Map inner updates to outer inputs/outputs
+        updates = {}
+        for rng, update in inner_updates.items():
+            inp_idx = op.inner_inputs.index(rng)
+            out_idx = op.inner_outputs.index(update)
+            updates[node.inputs[inp_idx]] = node.outputs[out_idx]
+        return updates
+
+
+class _CustomSymbolicDist(Distribution):
+
+    rv_type = CustomSymbolicDistRV
+
+    @classmethod
+    def dist(
+        cls,
+        *dist_params,
+        class_name: str,
+        random: Callable,
+        logp: Optional[Callable] = None,
+        logcdf: Optional[Callable] = None,
+        moment: Optional[Callable] = None,
+        ndim_supp: int = 0,
+        dtype: str = "floatX",
+        **kwargs,
+    ):
+        warnings.warn(
+            "CustomDist with symbolic random graph is still experimental. Expect bugs!",
+            UserWarning,
+        )
+
+        dist_params = [as_tensor_variable(param) for param in dist_params]
+
+        if logcdf is None:
+            logcdf = default_not_implemented(class_name, "logcdf")
+
+        if moment is None:
+            moment = functools.partial(
+                default_moment,
+                rv_name=class_name,
+                has_fallback=True,
+                ndim_supp=ndim_supp,
+            )
+
+        return super().dist(
+            dist_params,
+            class_name=class_name,
+            logp=logp,
+            logcdf=logcdf,
+            random=random,
+            moment=moment,
+            ndim_supp=ndim_supp,
+            **kwargs,
+        )
+
+    @classmethod
+    def rv_op(
+        cls,
+        *dist_params,
+        class_name: str,
+        random: Callable,
+        logp: Optional[Callable],
+        logcdf: Optional[Callable],
+        moment: Optional[Callable],
+        size=None,
+        ndim_supp: int,
+    ):
+        size = normalize_size_param(size)
+        dummy_size_param = size.type()
+        dummy_dist_params = [dist_param.type() for dist_param in dist_params]
+        dummy_rv = random(*dummy_dist_params, dummy_size_param)
```

Comment:
Can we block access to the `Model` context so that if users try to call `pm.Normal("x")` in the random function it errors out?

Reply:
Something like 

```python
with error_on_model_access():
  random(*dummy_dist_params, dummy_size_param)
```

Reply:
Implemented one approach in the last commit